### ISSUE TITLE
FR: Support for "iPad apps for Mac" aka Marzipan/Catalyst

### ISSUE THREAD

**Issue by siburb** [firebase/firebase-ios-sdk#3144]
### [REQUIRED] Step 2: Describe your environment

  * Xcode version: 11
  * Firebase Component: All

#### Steps to reproduce:

Attempting to build an iPad app for Mac, but it uses Firebase via Cocoapods, but we receive a lot of the following error:

`Building for UIKit for Mac, but the linked framework '<frameworkName>.framework' was built for <unknown>. You may need to restrict the platforms for which this framework should be linked in the target editor.`

<img width="462" alt="Screen Shot 2019-06-06 at 11 09 06 am" src="https://user-images.githubusercontent.com/1744275/59055910-cfb23c00-884b-11e9-9d96-990a669258ea.png">

Apparently, the Firebase binaries need to be built using Xcode 11 with the "Mac" option checked.

#### Relevant Code:

**Comment by ryanwilson** [comment]
Thanks for the report, @siburb! There are some docs and sessions we need to catch up on in order to find out how to get this working for folks. We're working on it and will update this issue as we find out more :)

**Comment by paulb777** [comment]
And, as always, PRs are welcome!

**Comment by EliteTechnicalCare** [comment]
Any update on this? I would love my iPad / Mac apps to all run on FB back end.

**Comment by morganchen12** [comment]
No update yet. In terms of infrastructure, adding Catalyst support would mean adding most of the support for AppKit apps as well. We're in the process of figuring that out and will have more to share soon.

**Comment by EliteTechnicalCare** [comment]
Are there any work arounds like building it for mac and the importing it into a project?

**Comment by morganchen12** [comment]
Theoretically what you're describing should be possible, since Catalyst apps are built on macOS apps, but will only work for the existing macOS-compatible libraries and defeats the purpose of Catalyst (at that point you're just building an AppKit app). You could then [create a bundle using Firebase and load it into your Catalyst app](https://www.highcaffeinecontent.com/blog/20190607-Beyond-the-Checkbox-with-Catalyst-and-AppKit).

**Comment by guidedways** [comment]
I really hope it's not going to only going to work for Catalyst supported apps. We really need this to work for all macOS apps with or without Catalyst.

**Comment by morganchen12** [comment]
AppKit support is also on our list of things to do. Fortunately, the Firebase website wouldn't need to distinguish between the two (i.e. console can display both as platform macOS), so much of the work is overlapping.

**Comment by appfrilans** [comment]
Is it possible to include firebase for macOS (UIKit for Mac) via pod 'FirebaseCore' ? I tried to make a separate target and set it in its own CocoaPods target (platform :macos, '10.15'), but importing via "import FirebaseCore" just gives me an error:

No such module 'FirebaseCore'

Any ideas? Or is this way impossible?

**Comment by ryanwilson** [comment]
@appfrilans as far as I can tell, UIKit for Mac apps need an XCFramework instead of regular frameworks (please someone jump in and correct me if this is incorrect 🙂) - CocoaPods doesn't currently support that. I'm installing Catalina beta now on and hope to get XCFrameworks built for the open source pods that currently support macOS, I'll be updating this thread as I discover things (may not be until early next week that I'm successful, but I'll try to get something ASAP).

**Comment by jonas-sleepcycle** [comment]
Here are someone who have gotten iOS frameworks working for UIKit for mac (not macOS though): 

https://forums.developer.apple.com/message/362577

**Comment by ryanwilson** [comment]
Thanks for the link, I tried a bunch last night but only ran into multiple errors with no real feedback, only an error message: `The operation couldn’t be completed. (XCBCore.XCFrameworkValidationError error 8.)`.

To add a snippet of information from the above link, Catalyst frameworks must be built against the iOS SDK not the macOS SDK.

We're currently relying on CocoaPods for the framework build during our zip file as well, still trying to figure out the best way to build for the new platform. There's very little documentation on XCFrameworks and the tool doesn't provide much information on failures - but we're trying and will continue to try 😄

**Comment by EliteTechnicalCare** [comment]
> Thanks for the link, I tried a bunch last night but only ran into multiple errors with no real feedback, only an error message: `The operation couldn’t be completed. (XCBCore.XCFrameworkValidationError error 8.)`.
> 
> To add a snippet of information from the above link, Catalyst frameworks must be built against the iOS SDK not the macOS SDK.
> 
> We're currently relying on CocoaPods for the framework build during our zip file as well, still trying to figure out the best way to build for the new platform. There's very little documentation on XCFrameworks and the tool doesn't provide much information on failures - but we're trying and will continue to try 😄

Does the new Xcode Beta provide a better error or the same useless nondescript ones?

**Comment by ryanwilson** [comment]
> Does the new Xcode Beta provide a better error or the same useless nondescript ones?

Unfortunately this was last night with the new beta.

**Comment by ryanwilson** [comment]
Good news: I was able to build some static `x86_64` libraries that should work with UIKit for Mac apps.

Bad news: The way we currently package the frameworks with `lipo` fails since there are two libraries with the same architecture (simulator is also `x86_64`).

I'll try to use the `-library` and `headers` parameters from `xcodebuild -create-xcframework` and see if I can package an XCFramework that works in a Catalyst app.

**Comment by dreampiggy** [comment]
> as far as I can tell, UIKit for Mac apps need an XCFramework instead of regular frameworks (please someone jump in and correct me if this is incorrect 🙂)

Is this true ? I can build a normal UIKit for macOS with `x86_64` architecture. The `.xcframework` is only one collection form of multiple individual `.framework`s.

**Comment by girishw** [comment]
Any updates on this? This would be critical for many apps.

**Comment by ryanwilson** [comment]
Still working away, apologies for the radio silence. 

I'll post when I can get something to work for `FirebaseCore` since that's what I'm starting with. I've been able to build a slice that works with UIKit for Mac, but currently am battling code-signing since it's now required for frameworks built for the new platform.

> > as far as I can tell, UIKit for Mac apps need an XCFramework instead of regular frameworks (please someone jump in and correct me if this is incorrect 🙂)

> Is this true ? I can build a normal UIKit for macOS with x86_64 architecture. The .xcframework is only one collection form of multiple individual .frameworks.

Sorry for the confusion, I was thinking more from a distribution standpoint. Yes one can make a standard `.framework` that works with UIKit for Mac apps, but in order to have a "fat library" that is one `.framework` that works with all platforms we'd need to switch to `.xcframeworks`. It's either that, or ship multiple `.framework`s for the same SDK and leave it up to the developer to add the ones with the right architecture to the right targets in their app - it's pretty messy and error-prone.

**Comment by ryanwilson** [comment]
As a side note - if anyone has successfully taken a CocoaPod and built it to support both UIKit for Mac as well as iOS + simulator, please share any resources you can 😄

**Comment by kaleidscpe** [comment]
Hi, has anyone taken a look at this solution? https://github.com/firebase/firebase-ios-sdk/issues/583#issuecomment-502055434

@ryanwilson

**Comment by ryanwilson** [comment]
@kaleidscpe must have missed that, will take a look when I can.

Otherwise I have an xcframework for FirebaseCore, GoogleUtilities, and FirebaseStorage although I'm having some difficulties getting Storage working - there's a `+load` that's not getting called which causes the Storage instance returned to be `nil`.

This is the `+load` call:
https://github.com/firebase/firebase-ios-sdk/blob/d8a0e5e61bfaf5f3a8f6ca93a356b39c1d94eb47/Firebase/Storage/FIRStorageComponent.m#L54

Adding `+ObjC` to my linker flags did not work. Still investigating.

**Comment by ryanwilson** [comment]
Turns out it was a silly project configuration error and I got it up and running. Once I verify things I'll probably create a GitHub release with the compiled XCFrameworks for testing purposes. Note: these will only be the open source frameworks to start (the ones available in this GitHub repo).

**Comment by jlubeck** [comment]
@ryanwilson any chance you can share your findings with the GoogleMaps devs? Looks like the same error is also happening with that SDK, but I can't find a Github repo for GoogleMaps to create a ticket...

**Comment by bbrv2018** [comment]
Good afternoon, we are super happy that you will allow us to use Firebase for our iPad app on the Mac. Any idea when the new frameworks will be released? We had to abstract all our call to Firebase from our App to get it working at the moment. 
Thank you very much!

**Comment by khatchik** [comment]
Hey @ryanwilson, I am working with Atlassian and trying to ship Jira's Mac App.
Do you know if this ticket will also include porting the crashlytics library? 
If the answer is no, do you know to whom should I raise a ticket for this?

Thanks mate!

**Comment by jlaws** [comment]
@ryanwilson you should take a look at how the Realm team got it working

https://github.com/realm/realm-cocoa/issues/6163

They recommend using a Swift Package for realm in place of cocoapods since they can live side by side.

**Comment by stevekellogg** [comment]
Just wondering..  Is there any ETA as to when we might have a build to start working with?  ALSO, will this have all of the capabilities of the current iOS build?

TIA

**Comment by steipete** [comment]
Adding my +1 here. We're also currently porting our apps and SDKs to Mac Catalyst, and had to mock Firebase. Happy to help if you're stuck, my DMs are open (@steipete on Twitter)

**Comment by girishw** [comment]
Same here, +1

**Comment by mylogon341** [comment]
+1 This is critical for us to be ready. We’ve worked around it up to now but need this 100%. Please respond, Firebase people 🙏🏻

**Comment by paulb777** [comment]
Thanks for the feedback.  Catalyst is a high priority for us, but its progress has been slowed by some urgent issues and summer vacations. We'll get a more concrete update out in the next two or three weeks. 

In the meantime, we're happy to look at PRs.

Has anyone else had success with the strategy at https://github.com/firebase/firebase-ios-sdk/issues/583#issuecomment-502055434?

**Comment by paulb777** [comment]
Related CocoaPods issue at https://github.com/CocoaPods/CocoaPods/issues/8877

**Comment by michael-mansour** [comment]
+1 this is also very critical for us 🙏🏼

**Comment by appfrilans** [comment]
+1 Very critical for us as well 👍

**Comment by kaleidscpe** [comment]
I hope to be able to work on my Mac app next month. +1

**Comment by Brian-Co** [comment]
+1 It’s very critical for us as well! 👍 Would really appreciate being able to work on Catalyst in the coming weeks 😃

**Comment by KevinMorais** [comment]
+1 also looking forward to implement Firebase (and Google Maps) with Catalyst 🙏🏼

**Comment by guidedways** [comment]
And I’m hoping plans of porting it to the Mac (normal macOS support without catalyst) are still there.

**Comment by EliteTechnicalCare** [comment]
Is there at least a rough timeline or a beta release so hopefully we can begin testing our apps and get our apps on the Mac App Store by launch day?

**Comment by paulb777** [comment]
Added a GitHub Catalyst label to help follow the progress on [issues](is:issue catalyst label:Catalyst) and [pull requests](is:pr catalyst label:Catalyst ).

**Comment by paulb777** [comment]
We're striving to have iOS SDK support for the open source pods by the time that Catalyst goes out of beta. Closed source pods and binary distribution support will likely take longer.

If you'd like to try it out, see the details below. Make sure your Podfile does not include any closed source pods like `FirebaseAnalytics`, `Firebase/Core`, `FirebasePerformance` or any FirebaseML.

**Introduction**
At WWDC 2019, Apple announced Catalyst. Catalyst adds UIKit support for macOS apps, which enables iOS apps to be targeted to the Mac with little more than a recompile.

After WWDC, Catalyst support became one of the most requested Firebase features.

**Closed Source Pods**
Adding support for the closed source pods - FirebaseAnalytics, FirebasePerformance, and the FirebaseML* pods - will require internal build infrastructure changes to support xcframeworks. It’s outside the scope of this document and will be dealt with in another document.

**Summary**
Here is the status of the open source Firebase libraries. Very little testing has been done beyond building, running unit tests, and some integration tests.  In the current master repo or pending PR's all are ready for early evaluation with the exception of FirebaseRemoteConfig, FirebaseInAppMessaging and FirebaseDynamicLinks.

- FirebaseCore
- FirebaseABTesting
- FirebaseAuth - Requires developers to the Keychain Sharing Capability to their build settings. Needs port of UIWebView to WKWebView  (#3549). After that, several deprecated build warnings should be addressed.
- FirebaseCoreDiagnostics
- FirebaseDatabase - Build fixed in #3512. Support will be published in Firebase 6.7.0.
- FirebaseDynamicLinks - Needs port of UIWebView to WKWebView. Framework dependencies need review. Is FDL useful on macOS?
- FirebaseFirestore - Requires signing workaround for gRPC resource.
- FirebaseFunctions - Functional with one known bug - #3514. Integration testing is blocked because npm is not yet available for Catalyst.
- FirebaseInAppMessaging - Builds but unit tests depend upon FirebaseAnalytics and DynamicLinks. Requires signing workaround for Resources.
- FirebaseInAppMessagingDisplay - Builds but unit tests depend upon XCUIDevice. Requires signing workaround for Resources.
- FirebaseInstallations - Functional, but requires developers to the Keychain Sharing Capability to their build settings.
- FirebaseInstanceID - Functional, but requires developers to the Keychain Sharing Capability to their build settings. Unit tests fixed in #3534.
- FirebaseMessaging -  Functional, but requires developers to the Keychain Sharing Capability to their build settings. 
- FirebaseRemoteConfig - Blocked on open sourcing, which is planned in the next month.
- FirebaseStorage
- GoogleDataTransport - Build requires updating deleted NSKeyedArchiver API changes(#3525).
- GoogleDataTransportCCTSupport
- GoogleUtilities - GULNetwork tests require Enabling "Incoming Connections" capability in the unit test host app. GULUserDefaults tests fix in #3544. `SecTrustEvaluate` deprecation should be addressed.

**Developing Firebase with Catalyst**
Install Catalina and Xcode 11.

For each podspec in https://github.com/firebase/firebase-ios-sdk, run

- pod gen {name here}.podspec --local-sources=./ --auto-open
- Check the Mac box in the App-iOS Build Settings
- Sign the App in the Settings Signing & Capabilities tab
- Click Pods in the Project Manager
- Add Signing to the iOS host app and unit test targets
- Select the Unit-unit scheme
- Run it to build and test

**Comment by filippozanfini** [comment]
How can I test Firebase with Catalyst support with my app?

**Comment by paulb777** [comment]
@filippozanfini If you restrict your pods to the ones listed above, add the pod to your app and build.  Some extra signing may be required as described above. 

If you run into problems, create another issue at this repo with a description of your configuration and we'll try to help.

**Comment by filippozanfini** [comment]
I didn't understand in which path I have to run "pod gen {name here}.podspec --local-sources=./ --auto-open"

**Comment by filippozanfini** [comment]
@paulb777 I don't understand in which path I have to run "pod gen {name here}.podspec --local-sources=./ --auto-open"

**Comment by paulb777** [comment]
@filippozanfini That's part of the instructions for developing Firebase itself.   From a client perspective, using Firebase should be just like using any other CocoaPod for building a Catalyst app.

**Comment by filippozanfini** [comment]
@paulb777 I updated the pods and now it works correctly. Thank you!

**Comment by pangia** [comment]
will this also work for manual 'non pod' implementations?

**Comment by guidedways** [comment]
And what about normal macOS support? Will this work?

**Comment by paulb777** [comment]
@pangia We've only verified Catalyst for source distributions which is only via CocoaPods or this GitHub repo.  We still need more investigation of setting up xcframeworks before setting up binary distributions.

@guidedways Normal macOS support is unchanged. See https://github.com/firebase/firebase-ios-sdk#macos-and-tvos

**Comment by girishw** [comment]
We're getting pretty close to iOS 13 & Catalina release. Probably a month away or less.

**Comment by steipete** [comment]
Ah, we're getting close to Catalina's release and neither `FirebasePerformance`, `FirebaseAnalytics` nor `FirebaseRemoteConfig ` are ready or have a clear timeline. That's unfortunate, looks like we'll have to ship with mocks.

Note that the new download format should be `xcframework`, to make things simple with same hw arch but different platform arch (mac vs catalyst). However CocoaPods might need some more time before https://github.com/CocoaPods/CocoaPods/issues/9148 is ready. (Same for Carthage, but there's a PR now)

If you're stuck on converting, [my DMs are open](https://twitter.com/steipete), we've had quite a ride bringing PSPDFKit to Catalyst on time, and doing the `xcframework` conversion.

**Comment by paulb777** [comment]
@steipete Thanks for creating the detailed CocoaPods `xcframework` issue!

As of yesterday's release, `FirebaseRemoteConfig` is now open sourced along with Catalyst support.

We still don't have a timeline for the closed source libraries. In addition to the `xcframework` challenges you've seen, we need to work through some challenges with Catalina and internal Google build systems.

**Comment by steipete** [comment]
Wohoo, congrats on open sourcing that one, that's a great step! Thanks!

**Comment by paulb777** [comment]
@Brian-Co As is detailed above, Firebase support is not complete and is only available as community support for the pods released as source from this repo.

Neither Firebase not CocoaPods yet have a solution for binary CocoaPods.

**Comment by jpv123** [comment]
Thanks @paulb777 for your answer. Do you have any possible ETA for this to be implemented? Unfortunately my knowledge is not so strong on this topics

**Comment by steipete** [comment]
Is Crashlytics open source? This would be our top priority to get into Catalyst in the next ~2 weeks.

**Comment by samedson** [comment]
Nothing to report on Crashlytics being open source at this time but we are aware of the excitement around this feature

**Comment by michael-mansour** [comment]
Hi, when I try to archive my project I get the following error with gRPC-Core:

> 'openssl_grpc/ssl.h' file not found

> 'openssl_grpc/bio.h' file not found

> 'openssl_grpc/bn.h' file not found

> 'openssl_grpc/rsa.h' file not found

> 'openssl_grpc/bio.h' file not found

**Comment by paulb777** [comment]
Investigating gRPC-Core archive issue at #4007

**Comment by steipete** [comment]
We shipped our Catalyst app (PDF Viewer) and we're *REALLY* missing Firebase Crashlytics. Is there a rough ETA for it? Is there anything we can do to help/beta-test?

**Comment by peterstojanowski** [comment]
@samedson Is there any update or ETA on Firebase Crashlytics SDKs? I assume this will replace old Fabric Crashlytics and be compatible with Catalyst?

**Comment by samedson** [comment]
Hey all, we are working towards providing this support - it's is a high priority for us. We'll let you know when there is news to share.

**Comment by samedson** [comment]
Right now we have community support for macOS here: https://github.com/firebase/firebase-ios-sdk#macos-and-tvos. More fully fledged support is TBD.

You can currently get Crashlytics on macOS without Analytics by setting it up as an iOS app in the Firebase Console, and including the pods:
```
pod 'Crashlytics'
pod 'FirebaseCore'
```

**Comment by fishandphil** [comment]
Good afternoon. 
We are trying to create the Firebase frameworks for our Catalyst App. Is it possible to have a bit more information about the following points:

1) "FirebaseFirestore - Requires signing workaround for gRPC resource."
What is the "workaround", were can we find information about what/how to do? 

2) When we do the last step: "Run it to build and test" 
What device should we do the "Build For Testing"?
- "Generic iOS Device" 
or 
- "My Mac"?
So far, we have been unable do build FirebaseCore on "My Mac" while it works fine on "Generic iOS Device". However, we were able to build FirebaseFirestore on "My Mac" which also build FirebaseCore I believe. Sadly, when integrating FirebaseFirestore for Catalyst in our project, we are getting about 30 "Undefined symbole" errors for grpc & leveldb. 
Unclear what we are doing wrong.

Thank you very much :)

**Comment by paulb777** [comment]
@fishandphil 

In general, you can search on the [Catalyst tag in this Issues list](https://github.com/firebase/firebase-ios-sdk/issues?q=is%3Aissue+is%3Aopen+label%3ACatalyst), to get more Catalyst-specific status.

Here are the answers:

1. See #3500. I updated the [comment above](https://github.com/firebase/firebase-ios-sdk/issues/3144#issuecomment-520530306) to clarify this. thanks!
2. "My Mac". There is a known open issue for archiving Firestore (#4007), but we haven't seen any issues in development builds.  If you're still having trouble with that, please open another issue with detailed repro instructions.

**Comment by girishw** [comment]
@paulb777  Need some clarifications on the current state and build instructions : 

- Can I install using the currently released Cocoapods (eg Firebase/Core (6.10.0)) or do I need to generate the pods from GitHub source? 
- After generating the pods from Github, do I simply reference the local pods from my Podfile? 
- Currently, when building from the released Cocoapods I get the following error : 

`ld: in ..../Documents/ReSchedule/Pods/FirebaseAnalytics/Frameworks/FIRAnalyticsConnector.framework/FIRAnalyticsConnector(FIRConnectorUtils_77ff1e12be6740765c87f1be0d421683.o), building for Mac Catalyst, but linking in object file built for iOS Simulator, file '.....Documents/ReSchedule/Pods/FirebaseAnalytics/Frameworks/FIRAnalyticsConnector.framework/FIRAnalyticsConnector' for architecture x86_64`

It looks like Firebase/Core (6.10.0) needs Firebase Analytics.

Thanks

**Comment by paulb777** [comment]
@girishw 

- You can use the released CocoaPods
- Use the pods just as you would for iOS
- As discussed above, FirebaseAnalytics is binary and has no support yet for Catalyst.  As you point out, Firebase/Core requires FirebaseAnalytics, so it cannot be in the Podfile. Separately, `Firebase/Core` is deprecated in favor of directly requesting the Firebase pods required.

**Comment by EliteTechnicalCare** [comment]
@paulb777 
I assume same issue with Crashalytics? Also is there anywhere we can follow the progress on that.

**Comment by paulb777** [comment]
@EliteTechnicalCare Only the [comment](https://github.com/firebase/firebase-ios-sdk/issues/3144#issuecomment-540622435) above.  Feel free to create another issue if you'd like to track Crashlytics for Catalyst specifically.

**Comment by Blackjacx** [comment]
This could be a game change: https://github.com/CocoaPods/CocoaPods/issues/9148

**Comment by steipete** [comment]
@Blackjacx We (PSPDFKit) did sponsor this particular CocoaPods feature. It will definitely make distribution easier in the future. It's not something that will help the Firebase team here, more additional work I would schedule after the initial support for Catalyst.

**Comment by ryanwilson** [comment]
It's definitely helpful for zip builds of our open source SDKs - huge thanks @steipete for sponsoring this feature (on behalf of Firebase as well as the whole community)!  There's still work to be done to get the closed source SDKs up and running with the internal build system.

**Comment by squm** [comment]
What is the current status, where I can check the progress on all this, not bothering anyone here at the issues ?

**Comment by paulb777** [comment]
@squm Good point, thanks! I updated the [repo's README](https://github.com/firebase/firebase-ios-sdk#tvos-macos-and-catalyst) today with a summary of the current state of Catalyst support. 

I'm going to close this umbrella issue in favor of more specific questions and issues.  Please check [existing Catalyst issues](https://github.com/firebase/firebase-ios-sdk/issues?q=is%3Aissue+is%3Aopen+label%3ACatalyst) and [create new issues](https://github.com/firebase/firebase-ios-sdk/issues/new) for additional issues and questions.

**Comment by girishw** [comment]
@paulb777 I am using GoogleSignIn with FirebaseAuth. Where does that fall in all this?

**Comment by paulb777** [comment]
GoogleSignIn is mostly out of scope for us here, but it's a [binary CocoaPod](https://github.com/CocoaPods/Specs/blob/master/Specs/d/4/0/GoogleSignIn/5.0.2/GoogleSignIn.podspec.json#L36) so Catalyst support would require more work. Probably best to follow up directly on one of the [GoogleSignIn channels](https://github.com/CocoaPods/Specs/blob/master/Specs/d/4/0/GoogleSignIn/5.0.2/GoogleSignIn.podspec.json#L36).

**Comment by pangia** [comment]
I don't see firebase analytics/crashlytics listed in the README. Are these not "_supported_" yet?

**Comment by girishw** [comment]
I found some requests here: https://github.com/googlesamples/google-services/issues/420

But radio silence.

**Comment by paulb777** [comment]
@pangia Follow #4316 for Crashlytics
@girishw I gave a Google-internal bump to the GSI team

**Comment by kajensen** [comment]
Added tickets for Catalyst support for Analytics (#4563) and RemoteConfig (#4564)

**Comment by girishw** [comment]
@paulb777 To integrate GoogleSignIn with Firebase Authentication on MacOS, can I use GTMAppAuth (https://github.com/google/GTMAppAuth) with the signIn(withCustomToken:completion:) method? 

Is there someplace else I should be asking this question?

**Comment by paulb777** [comment]
@girishw It's best to open a separate issue to more easily track and assign the question.
